crawl-ref/source/player.cc

```diff
@@ -5964,6 +5964,12 @@ int player::armour_class_with_specific_items(vector<const item_def *> items) con
 
     AC += sanguine_armour_bonus();
 
+    if (you.has_mutation(MUT_CURL)
+        && you.props[PALENTONGA_CURL_KEY].get_bool())
+    {
+        AC += 7 * scale;
+    }
+
```

Comment:
Refactoring idea: rather than toggling a bool and having this method check all bools, add/remove an "AC effector" instance on a vector. Calculating AC would iterate over all instances, get each AC component, and tally them up. This works because AC is additive.

Advantages:
 - You don't have to set `you.redraw_armour_class`; that can be done when pushing/popping an AC affector.
 - Avoids another long function that has lots of dependencies / does a lot of stuff. Keeps related code together.
 - More composable, more easily testable.

Possible disadvantages:
 - Can't see all AC calculation code together. Not sure how much this would matter.

Reply:
I do like the idea of refactoring! I've never liked the way that `redraw_armour_class` works.

One note is that `redraw_armour_class` is *also* responsible for updating AC colour *and* SH value & colour, but your suggestion shouldn't conflict with that, if I understand it correctly.